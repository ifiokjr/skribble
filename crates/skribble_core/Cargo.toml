[package]
name = "skribble_core"
version = { workspace = true }
authors = { workspace = true }
categories = { workspace = true }
documentation = "https://docs.rs/skribble_core"
edition = { workspace = true }
homepage = { workspace = true }
include = ["src/**/*.rs", "Cargo.toml", "readme.md"]
keywords = ["build", "atomic", "css"]
license = { workspace = true }
readme = "readme.md"
repository = { workspace = true }
rust-version = { workspace = true }
description = "Core features for the skribble atomic css library"

[dependencies]
derivative = { workspace = true }
derive_more = { workspace = true }
doc-comment = { workspace = true }
globset = { workspace = true }
heck = { workspace = true }
indent_write = { workspace = true }
indexmap = { features = ["serde"], workspace = true }
lazy_static = { workspace = true }
lightningcss = { workspace = true }
readonly = { workspace = true }
regex = { workspace = true }
serde = { workspace = true }
serde_json = { workspace = true }
skribble_color = { workspace = true, features = ["serde"] }
thiserror = { workspace = true }
typed-builder = { workspace = true }
vfs = { workspace = true }

[dev-dependencies]
insta = { features = ["json"], workspace = true }
