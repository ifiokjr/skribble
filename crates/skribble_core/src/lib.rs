#![deny(clippy::all)]

pub use error::*;

mod error;
