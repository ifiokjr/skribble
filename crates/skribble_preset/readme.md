# skribble_preset

<br />

[![Crate][crate-image]][crate-link] [![Docs][docs-image]][docs-link] [![Status][ci-status-image]][ci-status-link] [![Unlicense][unlicense-image]][unlicense-link]

## Installation

```toml
[dependencies]
skribble_preset = "0.0.0"
```

### Usage

```rust
use skribble_preset::*;
```

[crate-image]: https://img.shields.io/crates/v/skribble_preset.svg
[crate-link]: https://crates.io/crates/skribble_preset
[docs-image]: https://docs.rs/skribble_preset/badge.svg
[docs-link]: https://docs.rs/skribble_preset
[ci-status-image]: https://github.com/ifiokjr/skribble/workflows/ci/badge.svg
[ci-status-link]: https://github.com/ifiokjr/skribble/actions?query=workflow:ci
[unlicense-image]: https://img.shields.io/badge/license-Unlicence-blue.svg
[unlicense-link]: https://opensource.org/license/unlicense
