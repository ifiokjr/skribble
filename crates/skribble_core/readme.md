# skribble_core

> core library for skribble

<br />

[![Crate][crate-image]][crate-link] [![Docs][docs-image]][docs-link] [![Status][ci-status-image]][ci-status-link] [![Unlicense][unlicense-image]][unlicense-link]

## Installation

```toml
[dependencies]
skribble_core = "0.0.0"
```

### Usage

```rust
use skribble_core::*;
```

[crate-image]: https://img.shields.io/crates/v/skribble_core.svg
[crate-link]: https://crates.io/crates/skribble_core
[docs-image]: https://docs.rs/skribble_core/badge.svg
[docs-link]: https://docs.rs/skribble_core
[ci-status-image]: https://github.com/ifiokjr/skribble/workflows/ci/badge.svg
[ci-status-link]: https://github.com/ifiokjr/skribble/actions?query=workflow:ci
[unlicense-image]: https://img.shields.io/badge/license-Unlicence-blue.svg
[unlicense-link]: https://opensource.org/license/unlicense
