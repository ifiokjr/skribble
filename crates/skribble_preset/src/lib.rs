#![deny(clippy::all)]
#![forbid(clippy::indexing_slicing)]

doc_comment::doctest!("../readme.md");

use data::*;
use enums::*;
use serde::Deserialize;
use serde::Serialize;
use skribble_core::*;
use typed_builder::TypedBuilder;

pub(crate) mod data;
mod enums;

#[derive(Debug, Clone, Default, Deserialize, TypedBuilder, Serialize)]
pub struct PresetDefault {
  /// Choose the palette colors from either Tailwind or OpenColors.
  #[builder(default, setter(into))]
  pub palette: PaletteType,
  /// Choose how `light` and `dark` mode are handled. via `prefers-color-scheme`
  /// or `class` based.
  #[builder(default, setter(into))]
  pub dark_mode: DarkMode,
  /// Choose to ignore the color variables so that you can define your own
  /// custom theme.
  #[builder(default, setter(into))]
  pub ignore_colors: bool,
}

impl Plugin for PresetDefault {
  fn get_data(&self) -> PluginData {
    PluginData::builder()
      .id("skribble_preset")
      .name("Default Preset")
      .description(
        "This plugin provides a default preset for Skribble which is similar to `tailwind`, \
         `windi` and `unocss`.",
      )
      .build()
  }

  fn mutate_config(&mut self, config: &mut PluginConfig, _options: &Options) -> AnyEmptyResult {
    self.update_palette(&mut config.palette);
    self.update_media_queries(&mut config.media_queries);
    self.update_modifiers(&mut config.modifiers);
    self.update_variables(&mut config.variables);
    self.update_keyframes(&mut config.keyframes);
    self.update_atoms(&mut config.atoms);
    self.update_named_classes(&mut config.classes);
    self.update_value_sets(&mut config.value_sets);

    Ok(())
  }
}

impl PresetDefault {
  fn update_media_queries(&self, media_queries: &mut MediaQueries) {
    media_queries.extend(MEDIA_QUERIES.clone());

    if self.dark_mode == DarkMode::Media {
      media_queries.extend(DARK_MEDIA_QUERIES.clone());
    }
  }

  fn update_palette(&self, palette: &mut Palette) {
    palette.extend(self.palette.palette());
  }

  fn update_modifiers(&self, modifiers: &mut Modifiers) {
    modifiers.extend(MODIFIERS.clone());

    if self.dark_mode == DarkMode::Class {
      modifiers.extend_group(DARK_PARENT_MODIFIERS.clone());
    }
  }

  fn update_variables(&self, css_variables: &mut CssVariables) {
    if !self.ignore_colors {
      css_variables.extend(COLOR_CSS_VARIABLES.clone());
    }

    css_variables.extend(CSS_VARIABLES.clone());
  }

  fn update_keyframes(&self, keyframes: &mut Keyframes) {
    keyframes.extend(KEYFRAMES.clone());
  }

  fn update_atoms(&self, atoms: &mut Atoms) {
    atoms.extend(ATOMS.clone());
  }

  fn update_named_classes(&self, named_classes: &mut NamedClasses) {
    named_classes.extend(NAMED_CLASSES.clone());
  }

  fn update_value_sets(&self, value_sets: &mut ValueSets) {
    value_sets.extend(ATOM_VALUE_SETS.clone());
    value_sets.extend(ANIMATION_VALUE_SETS.clone());
  }
}

#[cfg(test)]
mod __tests;
