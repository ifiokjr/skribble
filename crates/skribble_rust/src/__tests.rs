use skribble_core::*;
use skribble_preset::PresetDefault;

use super::*;

#[test]
fn default_can_be_added_to_runner() {
  let default_preset = PresetDefault::builder().build();
  let rust_plugin = RustPlugin::builder().build();

  let config: StyleConfig = StyleConfig::builder()
    .plugins(vec![
      PluginContainer::from(default_preset),
      PluginContainer::from(rust_plugin),
    ])
    .build();

  let mut runner = SkribbleRunner::new(config);
  let _ = runner.initialize();
  let result = runner.generate().unwrap();
  let generated = result.get(0).unwrap();
  let content = &generated.content;
  insta::assert_display_snapshot!(content);
}
