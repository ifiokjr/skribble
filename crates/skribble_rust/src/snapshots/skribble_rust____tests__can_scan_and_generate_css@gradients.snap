---
source: crates/skribble_rust/src/__tests.rs
expression: scanned.code
---
@layer base, alias, default, priority-class;

@layer default {
  :root {
    --sk-gradient-from-position: ;
    --sk-gradient-via-position: ;
    --sk-gradient-to-position: ;
    --sk-gradient-stops: ;
    --sk-gradient-from: ;
    --sk-gradient-to: ;
  }

  .from-color\:\$cyan500 {
    --sk-gradient-from-position: ;
    --sk-gradient-via-position: ;
    --sk-gradient-to-position: ;
  }

  .to-color\:\$blue500 {
    --sk-gradient-from-position: ;
    --sk-gradient-via-position: ;
    --sk-gradient-to-position: ;
  }

  .bg-gradient\:\$to-right {
    background-image: linear-gradient(to right, var(--sk-gradient-stops));
  }

  .from-color\:\$cyan500 {
    --sk-gradient-from: #06b6d4 var(--sk-gradient-from-position);
    --sk-gradient-to: #06b6d400 var(--sk-gradient-to-position);
    --sk-gradient-stops: var(--sk-gradient-from), var(--sk-gradient-to);
  }

  .to-color\:\$blue500 {
    --sk-gradient-to: #3b82f6 var(--sk-gradient-to-position);
  }
}

